### ISSUE TITLE
Named type "Domain.Model" is invalid: Type string "Domain.Model" cannot be resolved

### ISSUE THREAD

**Issue by jan-johansson-mr** [dotnet/orleans#4795]
Hi,

I have this strange exception happening. It happens randomly, about 50% of the time, kind of sending goes fine, then sending again triggers exception and so on. I found similar [issues](https://github.com/dotnet/orleans/issues/1536) from 2016. Some of the issues in the context of generic types. However, the type I'm using is a plain C# (POCO) type.

Here is my scenario.

I want to share some of my data types (models) between solutions, so I create a NuGet package that I install wherever I need the model. I do the same with interfaces (e.g. different solutions implements client to the same grain in some silo).

I've read the documentation [Serialization and Writing Custom Serializers](http://dotnet.github.io/orleans/Documentation/Advanced-Concepts/Serialization.html) about serialization, but I can not see what I'm doing wrong. I do use "DataContract" attribute, and also the "Serializable" attribute (as the document instruct me to use the "Serializable" attribute on types).

It is not clear from the documentation how I use other serializers (I see that it's possible to use other serializers instead of the one out-of-the-box - as long as all other parts of the environment uses the same serializer).

But now I've run out of ideas.

Here is the model that fails to be serialized (about 50% of the time):

```
    [Serializable]
    [DataContract]
    public class Model
    {
        [DataMember]
        public string Title { get; set; }

        [DataMember]
        public string Subtitle { get; set; }

        [DataMember(EmitDefaultValue = false)]
        public DateTime Created { get; set; }

        [DataMember(EmitDefaultValue = false)]
        public string Context { get; set; }
    }
```
And here is the code fragment using the model:

```
        _clusterClient
		.GetGrain<IModelGrain>(id)
		.Update(model);
```
The `_clusterClient` is a singleton object, initialized once (connecting), and then used throughout. I do not restart my applications between failures. So the serialization seem to be failing somewhat randomly during runtime. I do however measure each time it fails (try-catch).

Since the `Microsoft.Orleans.OrleansCodeGenerator.Build` can't handle .NET Core version 2.1 (yet), all my solutions are locked to .NET Core version 2.0 (2.0.2). I'm using Orleans v2.0.3 and v2.0 throughout my (.NET Core) solutions.

I'd be happy to have some thoughts on why such a "simple" scenario fails. If it was consistent, like failing all the time, then I'd go with what the exception tells me, that the type can not be resolved (not found) and work with it. But it is this scenario of the serializer failing on and off, that I think is strange.

Thanks,
Jan

**Comment by ReubenBond** [comment]
Your type looks fine. Can you confirm that code is indeed being generated for the model class? You can search for the class in your obj directory and see the results in the generated .orleans.g.cs file.

Other than that, you can look at the build log output by adding <OrleansCodeGenLogLevel>Trace</OrleansCodeGenLogLevel> to your csproj files.

You could put your grains into .NET Standard 2.0 assemblies and run your silos on .NET Core 2.1.

I'm working on a long term fix for the issue with compatibility.

**Comment by ReubenBond** [comment]
The error that the type cannot be resolved indicates that there was no code generated for that type. I'm assuming that's the case. Perhaps it works in some calls and not in others because there is a mismatch between the client and silo. I.e, one has generated code and the other doesn't.

Which project is the code generator installed into? It should be on the models project so that all consumers can see the generated code.

I'm unsure about why it would work some of the time when you aren't restarting any of the hosts (and presumably all of your hosts are running the exact same builds). I'd need more detail about method signatures which succeed and fail to tell you that.

**Comment by jan-johansson-mr** [comment]
Thanks for your feedback.

I'm going to go through the solutions, and to try to make them as clean as possible (I have tried some ideas and right now its a mess). I checked for the .orleans.g.cs, and they are present in the grain projects, but not in the model project. I guess the code is generated where the model is used (the model project declares the models, but no code uses them - the generator is not installed in the model projects and the models get exported by NuGet:ing the models and then imported in other projects that uses the models).

One thing though, the model is exposed in a Web API, and then used in a call to a grain. I don't see why it should have an impact, but maybe it has.

BR,
Jan

**Comment by jan-johansson-mr** [comment]
A quick feedback for now.

Including the code generator with the models didn't help. In fact, I'm even worse of now than before. Another method started to fail (it gets a paginated collection via: `GetItems(int skip, int take)`, the function returns a simple POCO C# model with properties `int` and `Guid`), but here is the catch, the client can invoke the function, it hits the Silo grain and return the model okay, but then when the client side wants to unwrap the package, the exception happens.

I've found references to the models in the `.deps.json` on the client side, and indeed I can get the type with `Type.GetType("fullname, assemblyname")` method (the DLL resides with the NuGet repository, and is not in the bin directory of the executable - kind of GAC:ish - since I've installed the package with the model with the client).

So I guess the real reason that it's not working, is simply because of the missing `.orleans.g.cs` thing in the build of the client. When I installed the generator on the model side, the `.orleans.g.cs` are present in the model project. My conclusion so far is that the code generated in the model project is not present in the model binary (DLL) and therefore not present in the NuGet package of the model either. On the side note, when I used the code generator in the client project, it generated `.orleans.g.cs`, but with no reference to the models (since they're installed with NuGet and not present as source with the client - I guess). So doing this didn't help either.

I can be wrong in my conclusion of course. I'll continue to work on the issue and see if I can track down why I have the problems.

Thanks,
//Jan

**Comment by ReubenBond** [comment]
Could you try explicitly adding the models assembly to the application parts manager in ConfigureApplicationParts? It seems the default configuration is not loading it.

**Comment by jan-johansson-mr** [comment]
That did the trick. When adding the application parts (models) explicitly, then it works. At least `GetItems(int skip, int take)`. I'll dive into the rest to see if I can make the whole chain work without hiccups. By the way, now I have the code generator into three projects, the model, the interface and then the grain implementation. Thanks!

**Comment by jan-johansson-mr** [comment]
Hmmm... I'm back to where I started. About 50-60% of the invocations to `Update(model)` triggers the exception as I described in my initial post...

**Comment by jan-johansson-mr** [comment]
Here is the setup for each project: model, interface and service (grain) are all standard2.0, the silo and web application (client) are .NET Core 2.1. Each model, interface and service (grain) project are installed with the code generator. The dependency chain is structured as given by the Onion architecture, with Silo and web application as infrastructure layer.
I can switch out the serializer as a test, but I'm not sure how to do that, since the documentation is somewhat vague on how to do it. Can you give me a hint on how to use another serializer?

**Comment by jan-johansson-mr** [comment]
I've restructured the whole solution, no NuGet, all projects in the same solution, however, the problem persists. How can I see if the type is resolvable or not, before I invoke the grain that results in the exception (as described in my first post)?. Can it have to do something with threading issues?

**Comment by jan-johansson-mr** [comment]
Closed this thread to be continued on #4803.
